lib/src/main/java/io/ably/lib/transport/ConnectionManager.java

```diff
@@ -987,72 +987,70 @@ public void onAuthUpdated(final String token, final boolean waitForResponse) thr
+                try {
+                    ProtocolMessage msg = new ProtocolMessage(ProtocolMessage.Action.auth);
+                    msg.auth = new ProtocolMessage.AuthDetails(token);
+                    send(msg, false, null);
+                } catch (AblyException e) {
+                    /* The send failed. Close the transport; if a subsequent
+                     * reconnect succeeds, it will be with the new token. */
+                    Log.v(TAG, "onAuthUpdated: closing transport after send failure");
+                    transport.close();
+                }
+                break;
 
-                case connecting:
-                    /* Close the connecting transport. */
-                    Log.v(TAG, "onAuthUpdated: closing connecting transport");
-                    ErrorInfo disconnectError = new ErrorInfo("Aborting incomplete connection with superseded auth params", 503, 80003);
-                    requestState(new StateIndication(ConnectionState.disconnected, disconnectError, null, null));
-                    /* Start a new connection attempt. */
-                    connect();
-                    break;
+            case connecting:
+                /* Close the connecting transport. */
+                Log.v(TAG, "onAuthUpdated: closing connecting transport");
+                ErrorInfo disconnectError = new ErrorInfo("Aborting incomplete connection with superseded auth params", 503, 80003);
+                requestState(new StateIndication(ConnectionState.disconnected, disconnectError, null, null));
+                /* Start a new connection attempt. */
+                connect();
+                break;
 
-                default:
-                    /* Start a new connection attempt. */
-                    connect();
-                    break;
-            }
+            default:
+                /* Start a new connection attempt. */
+                connect();
+                break;
+        }
 
-            /* Wait for a currentState transition into anything other than connecting or
-             * disconnected in a background thread */
-            singleThreadExecutor.execute(() -> {
-                boolean waitingForConnected = true;
-                while (waitingForConnected) {
-                    final ErrorInfo reason = waiter.waitForChange();
-                    final ConnectionState connectionState = currentState.state;
-                    switch (connectionState) {
-                        case connected:
-                            authUpdateResult.onUpdate(true, null);
-                            Log.v(TAG, "onAuthUpdated: got connected");
-                            waitingForConnected = false;
-                            break;
+        /* Wait for a currentState transition into anything other than connecting or
+         * disconnected in a background thread */
+        singleThreadExecutor.execute(() -> {
+            final ConnectionWaiter waiter = new ConnectionWaiter();
```

Comment:
I think that we should not create the `waiter` here but rather at the start of this method, just like it was done before :thinking: My understanding is that as soon as the waiter is created it attaches a listener and starts observing the connection state. At the start of the method we're triggering some connection-related actions, so it might be that the connection state happens before this line of code is reached, therefore, the waiter will never get the connection change it's waiting for. WDYT?

Reply:
Yes, that's a good idea and is done as a part of https://github.com/ably/ably-java/pull/825/commits/ac9df421b69f78c1b4a68ff5b3b1baec2ba9d28f